qualtran/bloqs/controlled_state_preparation.py

```diff
@@ -0,0 +1,483 @@
+(rot\_reg) which is added into a phase gradient. Then phase kickback causes an overall offset of
+$e^{i2\pi x/2^b}$, where $x$ is the angle value loaded and $b$ the size of the rot\_reg. Below is an
+example for rot\_reg\_size=2.
+
+First there is the rot\_reg register with the value to be rotated (3 in this case) and the phase
+gradient
+
+$$
+|3\rangle(e^{2\pi i 0/4}|0\rangle + e^{2\pi i 1/4}|1\rangle +
+          e^{2\pi i 2/4}|2\rangle + e^{2\pi i 3/4}|3\rangle).
+$$
+
+Then the rot\_reg $|3\rangle$ register is added to the phase gradient and store the result in the
+phase gradient register
+
+$$
+|3\rangle(e^{2\pi i 0/4}|3\rangle + e^{2\pi i 1/4}|0\rangle +
+          e^{2\pi i 2/4}|1\rangle + e^{2\pi i 3/4}|2\rangle),
+$$
+
+but this is equivalent to the original state with a phase offset of $e^{2\pi i 1/4}$.
+
+
+References:
+    [Trading T-gates for dirty qubits in state preparation and unitary synthesis]
+    (https://arxiv.org/abs/1812.00954).
+        Low, Kliuchnikov, Schaeffer. 2018.
+
+"""
+
+from typing import Dict, Tuple
+
+import attrs
+import numpy as np
+from numpy.typing import ArrayLike
+
+from qualtran import (
+    Bloq,
+    bloq_example,
+    BloqBuilder,
+    BloqDocSpec,
+    SelectionRegister,
+    Signature,
+    SoquetT,
+)
+from qualtran.bloqs.arithmetic import Add
+from qualtran.bloqs.basic_gates import OneEffect, OneState, ZeroEffect, ZeroState
+from qualtran.bloqs.basic_gates.rotation import Rx
+from qualtran.bloqs.qrom import QROM
+from qualtran.bloqs.select_and_prepare import PrepareOracle
+
+
+@attrs.frozen
+class ControlledStatePreparationUsingRotations(PrepareOracle):
+    r"""Controlled state preparation without entangled residual using Ry and Rz rotations from [1].
+
+    Given a quantum state of which the list of coefficients $c_i$ is known
+    $$
+        |\psi \rangle = \sum_{i=0}^{N-1}c_{i}|i\rangle
+    $$
+    this gate prepares $|\psi\rangle$ from $|0\rangle$ conditioned by a control qubit
+    $$
+        U((|0\rangle + |1\rangle)|0\rangle) = |0\rangle |0\rangle + |1\rangle |\psi\rangle.
+    $$
+
+    Args:
+        state_bitsizes: number of qubits of the state.
+        rot_reg_bitsizes: size of the register that is used to store the rotation angles. Bigger values
+            increase the accuracy of the results.
+        state: tuple of length 2^state_bitsizes that contains the complex coefficients of the state.
+
+    References:
+        [Trading T-gates for dirty qubits in state preparation and unitary synthesis]
+        (https://arxiv.org/abs/1812.00954).
+            Low, Kliuchnikov, Schaeffer. 2018.
+    """
+
+    state_bitsizes: int
+    rot_reg_bitsizes: int
+    state: Tuple
+    uncompute: bool = False
+
+    @property
+    def selection_registers(self) -> Tuple[SelectionRegister, ...]:
+        return (
+            SelectionRegister(
+                "target_state", bitsize=self.state_bitsizes, iteration_length=self.state_bitsizes
```

Comment:
The iteration length should be `len(self.state)` (or `2 ** self.state_bitsize`)

Reply:
What does iteration length do exactly? Maybe I understood it wrong